Feature Unix Socket

Philipp66904 opened the pull request:
New PR for the huge changes

maf0115 committed:
Laid the groundwork for websocket communication

maf0115 committed:
Major bug and logic fallancies fixed

Philipp66904 commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
 
+DEFAULT_UNIX_SOCKET_PATH = "/tmp/thi_drone"
 
 class TelemetryNode(CommonNode):
```
As the telemetry node is not running on the drone (which I wrongly assumed last time), please don't derive from CommonNode.

di-math commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
+import select
 
+from rcl_interfaces.msg import Log
 from common_package_py.common_node import CommonNode
```
@Philipp66904 said this is not needed anymore. Can use generic, since node will not be located on the drone.

di-math commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
+                    print('connection accepted. Creating Subscription')
+
+                    '''
+                    Creating the subscription after the connection has accepted ensures that the callback function is executed only after the soccket is connected
```
Typo + make this one line comment

di-math commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
+                if sock == self.server_sock:
+                    # Accept the connection
+                    self.client_sock, _ = self.server_sock.accept()
+                    print('connection accepted. Creating Subscription')
```
Inconsistent capitalization

Philipp66904 commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
-        """
+        if os.path.exists(DEFAULT_UNIX_SOCKET_PATH): 
+            print('Socket already exists')
+            raise FileExistsError
```
It would be nice if you add your error message to the exception:
```suggestion
            raise FileExistsError("Socket already exists")
```

di-math commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
+    def subscription_callback(self, log_msg):
+        try: 
+            '''
+            Apparently, some characters in the log messages upset something in the json.loads(<received_data>.decode()) function. This is just a brute sanitization of the string, which needs to be implemented in a function of its own
```
Actually json.dump should already take care of this. As seen here https://docs.python.org/3/library/json.html#json.dump

Philipp66904 commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
-        super().__init__(id)
+        try:
+            # Bind server socket and listen for a connection. The node is blocked as long as a connection does not exist
+            self.server_sock.bind(DEFAULT_UNIX_SOCKET_PATH)
```
Add a print statement here so the user knows what is going on
```suggestion
            self.server_sock.bind(DEFAULT_UNIX_SOCKET_PATH)
            print("Blocking until connection is established.")
```

Philipp66904 commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
+
+            # Getting the reference for the client socket
+            readable, _, _, = select.select([self.server_sock], [], [])
+            for sock in readable:
```
What happens if there is nothing in readable?
You don't seem to capture this case.

Philipp66904 replied:
You could simply replace this whole for loop (including the if statement in it) with the following statement:
```suggestion
            if self.server_sock in readable:
```

di-math replied:
@maf0115, @Philipp66904 this code block is actually redundant. Just looked it up that listen is blocking if socket is not nonblocking. https://man7.org/linux/man-pages/man2/accept.2.html

We can actually throw it out.

Edit: my bad btw. wasn't aware of this, when I wrote the boilerplate code

maf0115 replied:
Just tried your change on my VM, it works. Will be applied and commited along other fixes

di-math commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
+
+        # In case of exceptions, close the socket and remove its path from disk
+        except Exception as e:
+            print(f'Error occurred: {e}')
+            self.server_sock.close()
```
Let's add this to the rospy.on_shutdown(h)

di-math commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
+        # In case of exceptions, close the socket and remove its path from disk
+        except Exception as e:
+            print(f'Error occurred: {e}')
+            self.server_sock.close()
+            os.remove(DEFAULT_UNIX_SOCKET_PATH)
```
Let's add this to the rospy.on_shutdown(h)

Philipp66904 commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
+            # Getting the reference for the client socket
+            readable, _, _, = select.select([self.server_sock], [], [])
+            for sock in readable:
+                if sock == self.server_sock:
```
What happens if you receive sockets in your `readable`, but none of them are equal to your `self.server_sock`?
You don't seem to capture this case.

Philipp66904 commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
+                    '''
+                    Creating the subscription after the connection has accepted ensures that the callback function is executed only after the soccket is connected
+                    '''
+                    self.subscription = self.create_subscription(
```
I would move the subscription creation to the end of the constructor and out of any if or whatever statement. -> Makes it easier to read
You're creating the subscription only when a connection is established and if you would capture all cases, the constructor will only reach it's end if a connection was established.

maf0115 replied:
Noted. Makses sense since we want the constructor to reach completion regardless of if a connection is established or not.

Philipp66904 commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
+            raise FileExistsError
+            
+        # Create server socket
+        self.server_sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
```
Why don't you include this into your `try` block?
Does this one never raise any exception?

maf0115 replied:
You're right, it slipped during development. Will be fixed

Philipp66904 commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
+                        Log,
+                        '/rosout',
+                        self.subscription_callback, 
+                        100)
```
You have to explicitly set the Quality of Service here.
100 means that ROS makes sure that every single message will reach it's destination (with a cache of 100 messages; TCP like).
This is not what we want: We want the telemetry node to receive the debug output in a UDP like style.
To achieve this, you have to set the Quality of Service accordingly.
You can either consult the docs or ask Johan on what you have to do here.

di-math commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
 
+DEFAULT_UNIX_SOCKET_PATH = "/tmp/thi_drone"
 
 class TelemetryNode(CommonNode):
```
@Philipp66904 said, this wouldn't be necessary to him, as the node isn't located on the drone

Philipp66904 replied:
In fact, I actually don't want @maf0115 to use it, as this will result in problems with the heartbeat (the Quality of Service is set wrong for the telemetry node and Mission Control would produce a lot of error messages as it doesn't expect any heartbeats from you).

Philipp66904 commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
+            '''
+            log_msg.msg = log_msg.msg.replace(':', '').replace('"', '').replace("'", "").replace("\{", "").replace("}", "")
+
+            self.client_sock.sendall(json.dumps({"type": "std" , "content" :  log_msg.msg}).encode())
```
Can you please provide me with examples of messages that were problematic?
You're doing a lot of sanitizing here, which on one hand will make literally all my log messages unreadable (e.g. `:` is removed completely) and on the other hand looks why too much for me.
Apart from the fact, that this is mostly taken care of by Python anyway.

maf0115 replied:
I am currently tryingt to figure out which messages cause this problem. Currently, when I run the node for testing, the error gets output by the listener program provided by @di-math when the program receives and tries to decode two messages right behind the other. I cannot really explain it that well yet, but I found consistent behavior using print statements in the terminal. Screenshot for reference: 

![image](https://github.com/THI-Drone/telemetry_package/assets/166122914/d45d6d85-ff96-46a5-8599-a0d1bcdb08d5)

![image](https://github.com/THI-Drone/telemetry_package/assets/166122914/f04c774b-7aac-4a03-97d9-f18e0ba8f8ba)

As you can see, the error gets output only when the receiver gets two messages on the same line (I'm sorry for the lack of a better explenation of the problem). I'll  be looking deeper into detail to understand why that's the case. I might be wrong and the examples I provided in the screenshots above are just due to some weird formatation in the terminal.

Philipp66904 replied:
I can tell you why your code is falling apart: You have two JSONs stitched together. The result is an invalid JSON which obviously breaks.

Philipp66904 commented on telemetry_package/main.py:
I'm missing a destructor here, that properly terminates the connection.
You should check if you need to end the unix socket or if it will end by itself.

J0hn-Lawrence commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
+            print('Socket already exists')
+            raise FileExistsError
+            
+        # Create server socket
```
Why not put this into an else branch? Would make it much more legible.

J0hn-Lawrence commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,74 @@
+                        '/rosout',
+                        self.subscription_callback, 
+                        100)
+                    self.subscription
```
what does this line do?

Philipp66904 replied:
He told me in the first PR, that this is to prevent warnings about an `unused variable`.

J0hn-Lawrence replied:
We should then add a comment with reference to documentation about this @maf0115

Philipp66904 reviewed (commented):
Thank you @maf0115 for a greatly improved version!

I have a few general things to add:
1. I would like you to add comments to your code (you can use GitHub Copilot or similar if you want to), to essentially add a brief summary to every function.
2. Make sure that you have in mind, that multiple drones might be in the air simultaneously. Your current approach doesn't seem to be able to handle this. Talk to @J0las to ask him about how this will be done (as far as I know, all drones share the same ROS network but separated into namespaces) and define an interface with @di-math that he knows from which drone / namespace the message is coming from.

J0hn-Lawrence commented on telemetry_package/main.py:
File completeley missing docstring comments for functions, if I am not mistaken,

J0hn-Lawrence reviewed (changes requested):
The scope of this PR is only to establish websocket connection to the Ground Station webserver, correct? Some changes (mainly documentation) necessary.

Philipp66904 commented:
@J0hn-Lawrence
Actually, they are using Unix Sockets for communication and not websockets.
You cannot really get it from the code, but you can clearly see it in the CDR presentation.

Unix Sockets communicate over shared memory and not the network: https://de.wikipedia.org/wiki/Unix_Domain_Socket

J0hn-Lawrence commented:
> @J0hn-Lawrence Actually, they are using Unix Sockets for communication and not websockets. You cannot really get it from the code, but you can clearly see it in the CDR presentation.
> 
> Unix Sockets communicate over shared memory and not the network: https://de.wikipedia.org/wiki/Unix_Domain_Socket

That's quirky, because the title of this PR is literally "Feature websocket" :laughing:. Nonetheless using unix sockets in this case is much preferred, as going over the network is unneccessary. Thanks for the clarification.

Philipp66904 commented:
@J0hn-Lawrence 
I thought that they were using websockets as well, because the branch is called that way.
But @di-math explained it to me.

maf0115 committed:
Create README.md

Added a detailed description of the current status and functionality of the program, as well as providing overview of the necessary fixes still to implement to ensure a clean functioning.

maf0115 committed:
Added comments for each function + first draft of sanitization

maf0115 committed:
Merge branch 'feature_websocket' of https://github.com/THI-Drone/telemetry_package into feature_websocket

di-math commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,73 @@
+            self.server_sock.listen(1)
+
+            # Getting the reference for the client socket
+            readable, _, _, = select.select([self.server_sock], [], [])
+            if self.server_sock in readable:            
```
```suggestion      
```
@maf0115 not needed

di-math commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,73 @@
+
+                
+            # Creating the subscription after the connection has accepted ensures that the callback function is executed only after the socket is connected           
+            self.subscription = self.create_subscription(
+                Log,
+                '/rosout',
+                self.subscription_callback, 
+                100)
+            self.subscription
```
Here we need to add a subscription to the heartbeat topic, to satisfy Philipp wish to see status of nodes

+ Add QoS Settings here

di-math commented on telemetry_package/main.py:
```diff
@@ -1,28 +1,73 @@
+ 
+    def subscription_callback(self, log_msg):
+        try: 
+            '''
+            Apparently, some characters in the log messages upset something in the json.loads(<received_data>.decode()) function. This is just a brute sanitization of the string, which needs to be implemented in a function of its own
+            '''
+            log_msg.msg = log_msg.msg.replace(':', '').replace('"', '').replace("'", "").replace("\{", "").replace("}", "")
+
+            self.client_sock.sendall(json.dumps({"type": "std" , "content" :  log_msg.msg}).encode())
```
```suggestion
            self.client_sock.sendall(json.dumps({"type": "std" , "content" :  log_msg.msg}).encode() + b"\x17")
```
@maf0115 this will fix the errors.

@Philipp66904 was right, pointing out, that the socket seems to be the issue, not the JSON. 

[See here](https://github.com/THI-Drone/Ground_Station/blob/dev/test/test_unix_socket/receiver.py), for adapted receiver dummy script.

di-math replied:
@maf0115 since node-library I use handles ends of readable streams it should be fine without this addition, this would just be for your debugging purposes.
In production please omit the ETB character.

maf0115 committed:
Added comments and other suggested measures

maf0115 committed:
Added subsctiption to heartbeat and updated QoS setting to match not deprecated settings

di-math commented:
Let's open up a new PR again...
